design/API_requirements_and_design.md

```diff
@@ -0,0 +1,59 @@
+# Thin-Edge API requirements and design
+
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
+* The tedge_api crate defines only messages related to the runtime as Shutdown.
+* The implementation of the mechanisms defined by the tedge_api are provided by
+  a tedge_runtime crate.
+* The final executable is build as an assemblage of all these plugin crates and
+  a `main.rs` that defines which plugins have to be instantiated and then
+  launched.
+
+---------
+
+# Non-functional Requirements
+
+- Plugins can be included in the final executable without having them 'instantiated'
+  - The motivation is to be able to deliver a battery included executable, ready
+    to be used for a large set of use-cases without any Rust expertise.
+  - Notably a battery-included executable should include an MQTT connection
+    plugin and a JSON plugin for telemetry messages.
+- Plugins are instantiated and configured through an external configuration at
+  runtime
+- At startup the configuration is read and from it a set of plugins 'to be
+  instantiated' is derived
+- One should be able to deduplicate parts of the configuration (to prevent
+  repetition)
+- Should plugin instances be changeable during runtime? (open question)
+- Should some plugin configurations be changeable at runtime or not?
+- Having an optional plugin handling SIGHUP and send an internal message
+  UpdateConfig to all the running plugin instances.
+- Would having multiple configuration files be an option? (open question)
+- Having plugins included but not instantiated should not impact other
+  instantiated plugins
+- A plugin should be able to signal the correctness of its configuration to the
+  best of its abilities
+- The runtime should not be able to start with a clear broken configuration
+- One plugin panicking should not impact other plugins or the runtime
+- If a plugin does not support a message that another plugin expects it to, the
+  runtime should not start
+  - Same thing for a missing required peer for a given kind, or two peers when
+    just one is expected.
+- Plugin/Application shutdown
+- Shutdown is signalled to all plugins, giving them a possibility to handle this
+  case
+- A plugin to handle SIGTERM? Or part of the core?
+- Plugins should be free to communicate with any other plugins, even in a
+  circular fashion.
```

Comment:
One aspect I am missing, and just now remembered is that communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

Reply:
To expand a bit:

Users do not necessarily get to choose which features a certain TE application has included. Having any form of 'implicit' communication (by for example sending it to the next best plugin that happens to accept the message type) makes using the application very unintuitive. 

Some example situations where this can happen:

 - Two plugins that happen to both respond to some SoftwareManagement messages but handle different overlapping-subsets of it. If you were to send a message that both are handling, you'd get duplicate messages and potentially duplicate actions which can be dangerous.
 - Enabling a plugin without keeping in mind _everything else_ in the configuration can potentially start causing effects on the device that the user did not intend to make, simply because now a target 'became available'

Reply:
>> Enabling a plugin without keeping in mind everything else in the configuration can potentially start causing effects on the >>  device that the user did not intend to make, simply because now a target 'became available'
>
> I don't understand this point.

I guess something like this: Plugins A, B and C are currently enabled and running. They work fine and properly. If now a plugin D gets enabled, and thin-edge now "figures out" a new way how these four plugins "should" communicate, simply enabling D might break the previously working setup of A, B and C.

That's why @TheNeikos proposes:

> [...] communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

---

On the other part of your comment:

> Some plugins might need to dispatch messages among a set of plugins delivering the same service but at different level/scope/target, this dispatch being done after a key extracted from the messages.

This sounds like the plugin that dispatches the messages already knows _when_ to dispatch these (lets call them) "special" messages to another plugin, right?
That would mean that the plugin would simply implement this logic: "If the message is special, send to plugin A, else to plugin B".

---

> On the other hand, other connections define the application and it makes little sense to change this connection in operations. For example a c8yplugin is expecting an mqtt_connection to c8y and a software_management_plugin.

I'm not entirely sure I understand your point here, could you maybe rephrase it?

Reply:
> I guess something like this: Plugins A, B and C are currently enabled and running. They work fine and properly. If now a plugin D gets enabled, and thin-edge now "figures out" a new way how these four plugins "should" communicate, simply enabling D might break the previously working setup of A, B and C.
That's why @TheNeikos proposes:
>> [...] communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

Okay, I got it. I fully agree that the system must not try to be smart (e.g. connecting plugins because the message types match).

However, I'm still not sure that this should be configured *only* at runtime (in some `config.toml`). I see cases where compile time configuration (in `main.rs`) would be more appropriate because there is no other meaningful alternatives.

I tried to give an example here:

>>On the other hand, other connections define the application and it makes little sense to change this connection in operations. For example a c8yplugin is expecting an mqtt_connection to c8y and a software_management_plugin.

> I'm not entirely sure I understand your point here, could you maybe rephrase it?

But was not clear enough. I think that some plugin connections, as between a `c8yplugin` and its `mqtt_connection` to c8y, there is no point to say again and again, in every runtime configuration and process launches, that this `c8y_mqtt_connection` instance is the mqtt connection of the `c8yplugin` instance.

Reply:
>> Some plugins might need to dispatch messages among a set of plugins delivering the same service but at different level/scope/target, this dispatch being done after a key extracted from the messages.

> This sounds like the plugin that dispatches the messages already knows when to dispatch these (lets call them) "special" messages to another plugin, right?
That would mean that the plugin would simply implement this logic: "If the message is special, send to plugin A, else to plugin B".

Definitely, the dispatcher plugin must implement the dispatching logic. However, this logic must be compatible with the actual plugin connections - especially when these ones are configured at runtime using plugin external names. Is the plugin named "A" in the config the correct target for the A messages?

Reply:
> However, I'm still not sure that this should be configured only at runtime (in some config.toml). I see cases where compile time configuration (in main.rs) would be more appropriate because there is no other meaningful alternatives.

Ah, I guess now I understand. You mean that someone might implement a usecase in more than one plugin, say (for the sake of an example), they implement their usecase in 3 plugins: A, B, C. One plugin (A) that actually talks to other plugins plus two "backend" plugins (B, C), that are only ever plugged into A. In this case, it would only make sense to provide _configuration_ for B and C in the `config.toml`, but never "wiring".

Does this represent what you mean?

Reply:
>> However, I'm still not sure that this should be configured only at runtime (in some config.toml). I see cases where compile time configuration (in main.rs) would be more appropriate because there is no other meaningful alternatives.

> Ah, I guess now I understand. You mean that someone might implement a usecase in more than one plugin, say (for the sake of an example), they implement their usecase in 3 plugins: A, B, C. One plugin (A) that actually talks to other plugins plus two "backend" plugins (B, C), that are only ever plugged into A. In this case, it would only make sense to provide configuration for B and C in the config.toml, but never "wiring".
Does this represent what you mean?

Yes. Though, said like this is sounds that it would be simpler to have a huge A-B-C plugin including the features of the 3 former ones.

Reply:
Exactly that would have been my next question :laughing: :+1: